### PR TITLE
Fix support for dialogs without response

### LINKED ISSUE #76: `main` branch no longer allows dialog-without-response
In v0.5.1, a dialog that does not have a response is valid:

```yaml
dialogues:
  - q: "*RST"
```

And when trying to `read()` after `write("*RST")` will result in a visa timeout error ('cause there's nothing in the read buffer).

```python
instr.write("*RST")
instr.query()  # raises VisaIOError timeout error
```

But in `main` (b0c2282), it's no longer valid and raises an error when parsing the yaml:

```
    def update_component(
        name: str, comp: Component, component_dict: Dict[str, Any]
    ) -> None:
        """Get a component from a component dict."""
        for dia in component_dict.get("dialogues", ()):
            try:
                comp.add_dialogue(*_get_pair(dia))
            except Exception as e:
                msg = "In device %s, malformed dialogue %s\n%r"
>               raise Exception(msg % (name, dia, e))
E               Exception: In device SMU, malformed dialogue {'q': '*RST'}
E               KeyError('r')
```

My main question is: Is this intentional?

It looks like the change happened in 970c18db8, specifically here: https://github.com/pyvisa/pyvisa-sim/commit/970c18db8f6513a66fda71e33fd9018a3f4e75d9#diff-00ae6e38b897e23f5d5f4ba58764e2de759b6e39ab2db9ea9c57d9315a20def4L62-R72

#### Yes it's intentional

How should we define a command that _shouldn't_ put anything in the read buffer?

The docs say to use a special `null_response` value, but that ends up actually getting returned:

```yaml
dialogues:
  - q: "*RST"
    r: null_response
```

```python
instr.write("*RST")
instr.query()  # expected: timeout error. Actual: "null_response"
```

Same with using `null` and the empty string `` in the yaml file:

```yaml
dialogues:
  - q: "*RST"
    r: null
```

```yaml
dialogues:
  - q: "*RST"
    r:
```

#### No, it's not intentional and is a bug

I'd be happy to add a test case for it and see if I can't resolve it.

### PR DESCRIPTION
**Pull request by dougthor42**
Fixes #76.

Re-add support for missing "response" key in the `dialogues` part and add tests for `_get_pair` and `_get_triplet` so that a similar regression won't happen in the future.

Note: Fully fixing the type annotations ended up being a much larger endeavor than expected, so I left that out of this PR and simply used `type: ignore`.

### CODE REVIEW

**Review by MatthieuDartiailh — COMMENTED** [review]
Out of curiosity, why can't you apply the same typing you use in the tests ?

**Comment by dougthor42** [comment]
**TL;DR:** You can see what all the type changes are in this branch: https://github.com/dougthor42/pyvisa-sim/tree/gh-76-fix-dialog-wo-response-types

Changing the return type of `_get_pair` from `str` to `OptionalStr` ends up cascading down to:

+ `channels.Channels.add_dialogue`
+ `channels.Channels._dialogues`
+ `channels.Channels._getters`
+ `component.Component.add_dialogue`
+ `component.Component._dialogues`
+ `component.Component._getters`
+ `component.Component._match_dialog`
   + which then requires some type guarding `if response is NoResponse:`
+ `devices.Device.add_eom`
+ `devices.Device._response_eom`
+ `devices.Device._eoms`

After doing all of that, you get to:

```
$ mypy pyvisa_sim/
pyvisa_sim/devices.py:258: error: Argument 1 to "extend" of "bytearray" has incompatible type "Union[bytes, Literal[Responses.NO]]"; expected "Iterable[SupportsIndex]"  [arg-type]
```

https://github.com/pyvisa/pyvisa-sim/blob/6fb7d9d01db76cc647e28f6aaddfcbe1c3b73e1f/pyvisa_sim/devices.py#L256-L258

Which requires it's own type guard `if eom is not NoResponse:`

```diff
                 if response is not NoResponse:
                     self._output_buffer.extend(response)
-                    self._output_buffer.extend(eom)
+                    if eom is not NoResponse:
+                       self._output_buffer.extend(eom)
```

I'm not yet familiar enough with the code to know if the type guard should be on both `self._output_buffer` (`if response is not NoResponse and eom is not NoResponse`) or just on one (like in the diff above). More testing and investigation is required.

In the end I didn't feel it was (a) worth it right now and (b) suitable to fix in this PR.

**Comment by MatthieuDartiailh** [comment]
Thanks. Type hints are a recent addition and I did not get them all right. I will merge as is but feel free to make a PR regarding teh type annotations and I will have a look.
